### PR TITLE
WIP: OMP-less JIT

### PR DESCRIPTION
**Pull request by henryiii**
This allows Clang-JIT to work, and OMP to be missing for JIT. (Clang + OMP not tested locally yet).

Adding all-in-one build and run of tests to allow travis to run less than 6 instances, so they can all run in parallel (5 parallel is limit, with no more than 2 of those on macOS). Hopefully the time is still reasonable.

I'm getting a clash of unsigned long vs. unsigned long long on macOS at the moment for JIT, otherwise the tests are okay. (And #227, as expected without patched Boost).

### CODE REVIEW

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -26,21 +26,42 @@ if(VEXCL_TEST_COVERAGE)
...
+        if(TARGET VexCL::CUDA)
+            _add_vexcl_test(${TEST_NAME}_cuda ${TEST_SOURCE} VexCL::CUDA)
+        endif()
+        if(TARGET VexCL::JIT)
+            _add_vexcl_test(${TEST_NAME}_jit ${TEST_SOURCE} VexCL::JIT)
+        endif()
+    else()
+        _add_vexcl_test(${TEST_NAME} ${TEST_SOURCE} VexCL::Backend)
+    endif()
+endfunction()
```
Comment (ddemidov) [inline tests/CMakeLists.txt]: Why not use `vexcl_add_executables` here (or in `_add_vexcl_test`) that you introduced in #225?

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -120,7 +141,9 @@ if (FFTW_FOUND)
 endif ()
```
Comment (ddemidov) [inline tests/CMakeLists.txt]: When `VEXCL_BACKEND=All`, target names for the fft test will be different. Using `vexcl_add_executables` should help here as well.

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -26,21 +26,42 @@ if(VEXCL_TEST_COVERAGE)
         )
 endif()
 
-function(add_vexcl_test TEST_NAME TEST_SOURCE)
+# Add one test with backend
+function(_add_vexcl_test TEST_NAME TEST_SOURCE BACKEND)
```
Comment (ddemidov) [inline tests/CMakeLists.txt]: This fails for `multiple_objects` target, which has more than one source file. Second source file gets picked for `BACKEND`.

File: tests/CMakeLists.txt (hunk 1 above)
Comment (henryiii) [reply tests/CMakeLists.txt]: Because we need to do an `add_test` on each one anyway.

File: tests/CMakeLists.txt (hunk 2 above)
Comment (henryiii) [reply tests/CMakeLists.txt]: Took a shot at fixing that when fixing the issue below.

File: tests/CMakeLists.txt (hunk 3 above)
Comment (henryiii) [reply tests/CMakeLists.txt]: That was a good idea; I also then could use `${ARGN}` to handle multiple sources correctly. :)

Comment (henryiii) [reply tests/CMakeLists.txt]: My first thought was to add keyword arguments, but this is better.

File: tests/CMakeLists.txt (hunk 1 above)
Comment (henryiii) [reply tests/CMakeLists.txt]: Also was trying to leave the boost unit def line PRIVATE, wasn’t sure if that was needed, mostly familiar with google test and catch.

**Comment by henryiii** [comment]
Would you mind dropping support for CMake < 3.8 for CUDA targets? CUDA is hard to get right before 3.8, things like `target_link_libraries` don't work properly with FindCUDA's specialized `cuda_add*` functions, but should work much better with the CUDA as a first-class language support. And Windows + CUDA support is pretty poor on Windows before CMake 3.9. I may be able to do it more generally, but just in case, would that be okay?

**Comment by ddemidov** [comment]
If `cuda_add_*` targets are the only problem here, then we can just wrap those in a conditional checking for cmake version. The only affected example in VexCL is `thrust-sort`, which serves as an example with interoperation with normal CUDA programs.

**Comment by henryiii** [comment]
Oh, didn't realize that CUDA is added with normal `add_executable`. It doesn't need NVCC? Then yes, that's only one example, which we can easily make version dependent.

**Comment by ddemidov** [comment]
Yes, VexCL uses driver API for the CUDA backend. NVCC is required, but only during runtime (to compile the generated kernels).

**Comment by ddemidov** [comment]
I like the idea, but this makes Travis-CI jobs to timeout. May be it is better to keep the `All` target (it makes local testing convenient), but leave Travis backends separate.

**Comment by henryiii** [comment]
That would be fine. It was somewhat experimental to try it on Travis. I just upgraded to CUDA 9 and that broke Eigen, so I’ve been busy preparing a patch for that, will try to get back to VexCL soon.

**Comment by henryiii** [comment]
And since it’s just one example in CUDA, it should be possible to manually set it up.

**Comment by henryiii** [comment]
Looks good, all required tests have passed. I'll work on CUDA improvements/fixes later and make a new PR for that.

**Comment by ddemidov** [comment]
Thanks again!

**Comment by henryiii** [comment]
Just tried master. CUDA build seems to be fine now.

**Pull request merged**